### ISSUE TITLE
Flexbox Layout

### ISSUE THREAD

**Issue by kaydwithers** [inuitcss/inuitcss#253]
WIth Bootstrap 4 coming out with an optional Flexbox grid. Can we also implement something similar?

http://v4-alpha.getbootstrap.com/layout/flexbox-grid/

**Comment by davidbwaters** [comment]
There's another Harry-inspired ITCSS project I found, Nebula-CSS, that has a nice Inuit-esque implementation:
[https://github.com/rbrtsmith/nebula-css/blob/master/nebula-css/objects/_grid.scss](url)
That's pretty close to what I'd like to see if in an Inuit flexbox grid.

Side note: I've found you can add `display: flex;` to `o-layout` and `flex-wrap: wrap;` to `o-layout__item`, you can have progressive equal height columns on browsers that support it (IE10+ I think) that fall back to the default display for browsers that don't.  I'm pretty sure that's all that's needed.  I can double check, if anyone's interested.

**Comment by kaydwithers** [comment]
A big reason i'd like a flex layout is to get rid of whitespace comments. i.e

```
   <div></div><!-- 

--><div></div>
```

It caused a lot of confusion at my work. One of our developers did not understand what they were and found and replaced them. And a back end developer also did the same. 

Our living styleguide also uses html-minifier and comments get removed. 

Developer's that join our project that are used to using Bootstrap or Foundation think it's odd. And it takes them a little while to understand the point of whitespace comments and how powerful Inuit layout can be.

**Comment by thewinger** [comment]
@kaydwithers in the last version of inuitcss/_objects.layout.scss you don't need to use the  trick
```css
53 // By default we use the `font-size: 0;` trick to remove whitespace between
54 // items.
```

**Comment by kaydwithers** [comment]
Ah awesome thanks, sorry I didn't see that. I would still really like a Flexbox layout. Is it something we are ignoring because we think the inline-block layout is the best grid solution?

**Comment by davidbwaters** [comment]
Probably a browser-support issue.  As a base for potentially large UI projects, many projects will need to support browsers before IE10.  Even though there's a usable flexbox polyfill, though, As a project aimed at being opinionated, I doubt Inuit would even want to make assumptions as to whether or not JS is enabled, so the polyfill approach wouldn't work.

I assume this is something they will point to taking a separately-maintained plugin approach with.  Since the width utilities work with any grid system, creating a flexbox based layout object wouldn't be too difficult.

Also, the inline-block approach is pretty darn flexible and offers some of the benefits of flexbox when compared to a float based grid.

**Comment by csshugs** [comment]
@kaydwithers
First of all: **We are not Bootstrap!** 😜 

Second: By default, inuitcss [now uses the `font-size: 0;`](https://github.com/inuitcss/inuitcss/blob/04abcb2cb1285a3223b23f3ef28790222e9ef906/objects/_objects.layout.scss#L53-L56) trick, to prevent you from needing to put comments in your markup.

But now to my most important question: Why do we want a flexbox-based grid-system? What does this offer us, what our current grid-system can't give us? 
I think our `inline-block` grid-system is very powerful. We don't have to clear floats, we can push and pull items. The only thing I miss sometimes is having equal height items, for which @davidbwaters [proposed a solution](https://github.com/inuitcss/inuitcss/issues/253#issuecomment-269081694) and I also [have worked out a "add-on" solution](http://codepen.io/csshugs/pen/OVovvo) for inuitcss via a modifier class.

For me, there isn't really a need to change to a flexbox-based grid-system. Flexbox is great and we all should thank the CSS-god on our knees every evening that we can now use this in production. But that doesn't mean we have to use it everywhere.

And of course, since we are **supporting IE9**, we can't provide a flexbox-grid by default.

**Comment by nicoqh** [comment]
While I agree with @csshugs, we've been using another grid system (using flexbox) for a while.

I just released it on npm so that I don't need to copy it over for every project, and I would be thrilled if anyone else would want to use it. I'm also very open to improvements. Whatever else, it's fun to finally have released something even though it's "yet another grid system" :)

**Comment by kaydwithers** [comment]
@csshugs did you take a look at how Bootstrap 4's [Flex grid](https://github.com/twbs/bootstrap/blob/v4-dev/scss/mixins/_grid.scss) works? It's an optional flex grid and it's a simple implementation.

By having a flex layout it makes Inuit appear a lot more modern which would appeal to a lot of users. Flexbox is amazing to use and it doesn't just have to be used for grids, it has a lot of other uses.

I love Inline-block grid, but it feels a little hacky. Using Flexbox is more of a best practice now.

According to Caniuse [IE9 usage](http://caniuse.com/usage-table) is now at 0.27%. Is IE9 something we want to continue to support? Even Bootstrap, a major framework is no longer supporting it. Do you think that by supporting IE9 and not using Flex, it makes Inuit look a little outdated?

**Comment by csshugs** [comment]
@nicoqh 
> https://github.com/nicoqh/inuit-flexgrid

❤️  That's exactly how we imagined our inuitcss ecosystem to be. We'd love to see more plugins arise in that exact manner. You had something missing in inuitcss, closed that gap with your own plugin and now, everyone can benefit from that plugin publicly available. Awesome 👍 

---

@kaydwithers 
> it makes Inuit appear a lot more modern which would appeal to a lot of users

I'd argue that inuitcss' job is to be the hottest CSS framework out there, pretending to be super cool and modern. We want to provide a robust framework that doesn't get in your way, making your life as easy as possible. What does it matter *how* (i.e. with what kind of CSS technique) we achieve the desired behavior.  The markup will be the same after all, no matter if we use the `inline-block` method or the `display: flex` method. If there is no significant advantage of the flexbox method over the inline-block method, I still don't see the point of compromising the IE9 support. Because...

> Do you think that by supporting IE9 and not using Flex, it makes Inuit look a little outdated?

...for me, that is rather a benefit than a handicap. Although it's really not inuitcss' primary aim to attract as many devs as possible, there may be devs who need to support IE9 in their projects and now cannot use Bootstrap anymore. Isn't it great, that they can still use inuitcss?
And honestly, appearing to be outdated isn't a point for me at all. The important question for me is: "Does it work as expected?" Any dev who uses something just because it's the newest shit probably shouldn't use inuitcss.

> Using Flexbox is more of a best practice now.

It's best practice for those who don't need to support IE9/10 anymore. It doesn't matter in the first place what the usage value in percentage is. If your client wants IE9 support than you have to deal with it. And here is my point again: We don't have to compromise this decision. By providing our current layout-system, there is no problem at all for supporting IE9. But if we decide to drop the inline-block method and provide the flexbox solution, there is no grid for IE9 supporters anymore.

We made the decision internally to support IE9 for the inuitcss version 6.0.0 for this exact reason. So in inuitcss core, we will not drop our current grid-system for a flexbox method. That doesn't mean, we are not willing to extend this functionality via a plugin later—There are already some out there, as you can see above. And of course we are considering this in any future inuitcss release.

**Comment by kaydwithers** [comment]
Okay understood. I'll stick to using a plugin. Hopefully we can use more flex in Inuit 7.

**Comment by csshugs** [comment]
> Hopefully we can use more flex in Inuit 7.

Agree. 

To make my point clear conclusively before I close this issue: It's all about setting a common denominator. As long as a flexbox-based solution does not provide us significant extra functionality over our current inline-block method, I don't see why we should lock out everyone who needs to support IE9 in their projects. I'm absolutely open for progressive enhancements like it's suggested in #265 though. And I just checked [the plugin](https://github.com/nicoqh/inuit-flexgrid) of @nicoqh again and I really have to say that this is exactly how we imagine our ecosystem to be. I love it!

**Comment by coryarmbrecht** [comment]
TL;DR Inuit Flex support would be great in my work. I'd like to use it everywhere.

Not supporting Flex does not benefit me in trying to support other platforms (that don't support CSS, but Flex). I'd like to have a shared codebase between web and native platforms, so that kind of makes me not want to use Inuit in these cases. I've always loved and tried to use Inuit where I can, so I'll just have to follow where @nicoqh 's plugin is going. Sounds like it might be straightforward to sort out.

Flex support might not make sense for InuitCSS [and web support], but the concepts/way of Inuit (IT...) for modern, emerging Flex, and native platforms would be awesome. I get it, and I could be totally off, but that's my thoughts! ^_^

**Comment by nicoqh** [comment]
@coryarmbrecht I'm happily accepting feature requests (or, even better, pull requests) for [inuit-flexgrid](https://github.com/nicoqh/inuit-flexgrid). I know there are certain flexbox features that the plugin could benefit from (e.g. ordering).
